File: CRM-Server/app/schemas/role.py
from __future__ import annotations
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from app.schemas.permission import PermissionResponse


class RoleBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=50, description="角色名称")
    code: str = Field(..., min_length=1, max_length=50, description="角色代码")
    description: Optional[str] = Field(None, description="角色描述")


class RoleCreate(RoleBase):
    pass


class RoleUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=50, description="角色名称")
    description: Optional[str] = Field(None, description="角色描述")


class RoleResponse(RoleBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class RoleWithPermissions(RoleResponse):
    permissions: List[PermissionResponse] = []


class UserRoleCreate(BaseModel):
    user_id: int = Field(..., description="用户ID")
    role_id: int = Field(..., description="角色ID")
    team_id: int = Field(..., description="团队ID")


class UserRoleResponse(BaseModel):
    id: int
    user_id: int
    role_id: int
    created_at: datetime

    class Config:
        from_attributes = True


class RolePermissionsUpdate(BaseModel):
    """批量更新角色权限的请求"""
    permission_ids: List[int] = Field(..., description="权限ID列表")

File: CRM-Client/src/api/role.ts
import request from '@/utils/request'
import {
  PermissionResponseSchema,
  RoleMutationResponseSchema,
  RoleResponseSchema,
  RoleUserResponseSchema,
  RoleWithPermissionsSchema,
  type PermissionResponse,
  type RoleMutationResponse,
  type RoleResponse,
  type RoleUserResponse,
  type RoleWithPermissions,
} from '@/schemas/role'

export type {
  PermissionResponse,
  RoleResponse,
  RoleUserResponse,
  RoleWithPermissions,
} from '@/schemas/role'

export interface RoleCreate {
  code: string
  name: string
  description?: string | null
}

export interface RoleUpdate {
  role_name?: string | null
  description?: string | null
}

export interface UserRoleCreate {
  user_id: number
  role_id: number
  team_id: number
}

export interface RoleQueryParams {
  skip?: number
  limit?: number
}

export interface RolePermissionsUpdate {
  permission_ids: number[]
}

const roleApi = {
  getRoles: async (params?: RoleQueryParams): Promise<RoleResponse[]> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.get('/v1/roles', { params })
    return RoleResponseSchema.array().parse(response)
  },

  getRole: async (roleId: number): Promise<RoleWithPermissions> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.get(`/v1/roles/${roleId}`)
    return RoleWithPermissionsSchema.parse(response)
  },

  createRole: async (data: RoleCreate): Promise<RoleResponse> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.post('/v1/roles', data)
    return RoleResponseSchema.parse(response)
  },

  updateRole: async (roleId: number, data: RoleUpdate): Promise<RoleResponse> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.put(`/v1/roles/${roleId}`, data)
    return RoleResponseSchema.parse(response)
  },

  deleteRole: async (roleId: number): Promise<RoleResponse> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.delete(`/v1/roles/${roleId}`)
    return RoleResponseSchema.parse(response)
  },

  assignRoleToUser: async (roleId: number, userId: number, teamId: number): Promise<RoleMutationResponse> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.post(`/v1/roles/${roleId}/users`, { user_id: userId, role_id: roleId, team_id: teamId })
    return RoleMutationResponseSchema.parse(response)
  },

  removeRoleFromUser: async (roleId: number, userId: number, teamId: number): Promise<RoleMutationResponse> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.delete(`/v1/roles/${roleId}/users/${userId}`, { params: { team_id: teamId } })
    return RoleMutationResponseSchema.parse(response)
  },

  getRoleUsers: async (roleId: number): Promise<RoleUserResponse[]> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.get(`/v1/roles/${roleId}/users`)
    return RoleUserResponseSchema.array().parse(response)
  },

  updateRolePermissions: async (roleId: number, permissionIds: number[]): Promise<PermissionResponse[]> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response: unknown = await request.put(`/v1/roles/${roleId}/permissions`, {
      permission_ids: permissionIds
    })
    return PermissionResponseSchema.array().parse(response)
  }
}

export default roleApi
